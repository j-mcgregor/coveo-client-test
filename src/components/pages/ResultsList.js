import React from 'react';
import { Link } from 'react-router-dom';
import S from '../../static/styles';
import Spinner from '../shared/Spinner';

const stringToKey = string => string.trim().replace(/\s/g, '-');

const ResultsList = ({ searchResults, loading }) => {
  const results = searchResults && searchResults.map(result => (
    <dt key={stringToKey(result.title)}>
      <Link to={`/results/${result.id}`}>
        {result.title}
      </Link>
    </dt>
  ));

  return (
    <S.ResultsList>
      {loading ? (<Spinner />) : (<dl>{results}</dl>)}
    </S.ResultsList>
  );
};

export default ResultsList;
