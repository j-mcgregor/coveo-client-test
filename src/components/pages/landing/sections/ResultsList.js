/* eslint-disable operator-linebreak */
import { get } from 'lodash';
import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';
import S from '../../../../static/styles';
import stringToKey from '../../../../utils/createKey';
import Spinner from '../../../shared/Spinner';
import StarRating from '../../../shared/StarRating';

const ResultsList = ({ searchResults, loading, selectProduct }) => {
  const results = searchResults.map((result, i) => {
    /* <<<<<<< LODASH GET IS CLEANER TO GET REQUIRED FIELDS >>>>>>> */
    const price = get(result, 'raw.tpprixnormal');
    const id = get(result, 'raw.tpproductid');
    const ml = get(result, 'raw.tpformat');
    const category = get(result, 'raw.tpcategorie');

    return (
      <Link
        to={`/products/${id}`}
        className="list-item columns"
        key={stringToKey(result.title, i)}
        onClick={() => selectProduct(result)}
      >
        <div className="column col-xs-6">
          <div className="columns">
            <div className="column col-12">{result.title}</div>
            <div className="column col-12">
              <small>{category}</small>
            </div>
          </div>
        </div>
        <div className="column col-xs-6 text-right">
          <div className="columns">
            <div className="column col-12">
              <StarRating percentScore={result.percentScore} />
            </div>
            <div className="column col-12">
              <small>
                {price} / {ml}
              </small>
            </div>
          </div>
        </div>
      </Link>
    );
  });

  return (
    <S.ResultsList>
      {loading ? <Spinner /> : <div className="container">{results}</div>}
    </S.ResultsList>
  );
};

ResultsList.propTypes = {
  searchResults: PropTypes.arrayOf(PropTypes.shape({})),
  loading: PropTypes.bool,
  selectProduct: PropTypes.func
};

ResultsList.defaultProps = {
  searchResults: [],
  loading: false,
  selectProduct: () => {}
};

export default ResultsList;
