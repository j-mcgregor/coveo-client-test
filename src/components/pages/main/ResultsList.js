/* eslint-disable operator-linebreak */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import S from '../../../static/styles';
import Spinner from '../../shared/Spinner';
import StarRating from '../../shared/StarRating';
import stringToKey from '../../../utils/createKey';

const ResultsList = ({ searchResults, loading, selectProduct }) => {
  const results =
    searchResults &&
    searchResults.map((result, i) => (
      <Link
        className="list-item columns"
        to={`/products/${stringToKey(result.title, i)}`}
        key={stringToKey(result.title, i)}
        onClick={() => selectProduct(result)}
      >
        <div className="column col-xs-12">{result.title}</div>
        <div className="column col-xs-12 text-right">
          <StarRating percentScore={result.percentScore} />
        </div>
      </Link>
    ));

  return (
    <S.ResultsList>
      {loading ? <Spinner /> : <div className="container">{results}</div>}
    </S.ResultsList>
  );
};

ResultsList.propTypes = {
  searchResults: PropTypes.arrayOf(PropTypes.shape({})),
  loading: PropTypes.bool,
  selectProduct: PropTypes.func
};

ResultsList.defaultProps = {
  searchResults: [],
  loading: false,
  selectProduct: () => {}
};

export default ResultsList;
