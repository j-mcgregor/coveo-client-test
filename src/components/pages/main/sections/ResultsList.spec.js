import '@testing-library/jest-dom/extend-expect';
import { mount, shallow } from 'enzyme';
import React from 'react';
import ResultsList from './ResultsList';

describe('ResultsList', () => {
  let defaultProps;

  beforeEach(() => {
    defaultProps = {
      searchResults: [],
      selectProduct: jest.fn(),
      loading: false
    };
  });

  it('should render and the items if present', () => {
    const props = {
      ...defaultProps,
      searchResults: [
        {
          title: 'Some title',
          raw: {
            tpprixnormal: 1.5,
            tpproductid: '123',
            tpformat: '330ml',
            tpcategorie: 'beer'
          },
          percentScore: 12.34
        }
      ]
    };
    const wrapper = shallow(<ResultsList {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render and empty container if searchResults is empty', () => {
    const wrapper = shallow(<ResultsList {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the spinner if loading', () => {
    const props = {
      ...defaultProps,
      loading: true
    };

    const wrapper = mount(<ResultsList {...props} />);

    expect(wrapper.find('svg')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger the selectProduct function when clicked', () => {
    const props = {
      ...defaultProps,
      searchResults: [
        {
          title: 'Some title',
          raw: {
            tpprixnormal: 1.5,
            tpproductid: '123',
            tpformat: '330ml',
            tpcategorie: 'beer'
          },
          percentScore: 12.34
        }
      ]
    };
    const wrapper = shallow(<ResultsList {...props} />);
    wrapper.find('Link').simulate('click');

    expect(props.selectProduct).toHaveBeenCalled();
    expect(props.selectProduct).toHaveBeenCalledWith({
      percentScore: 12.34,
      raw: {
        tpcategorie: 'beer',
        tpformat: '330ml',
        tpprixnormal: 1.5,
        tpproductid: '123'
      },
      title: 'Some title'
    });
  });
});
