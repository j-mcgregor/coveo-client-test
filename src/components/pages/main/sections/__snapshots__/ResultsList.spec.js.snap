// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsList should render and empty container if searchResults is empty 1`] = `ShallowWrapper {}`;

exports[`ResultsList should render and the items if present 1`] = `ShallowWrapper {}`;

exports[`ResultsList should render the spinner if loading 1`] = `ReactWrapper {}`;
